### PR TITLE
Docker autoconfig as a regular build rule

### PR DESCRIPTION
**Pull request by nlopezgi**
Docker autoconfig was a weird rule.
Needed to run with bazel run, outputs were produced in a directory outside of bazel runfiles tree. Other rules (e.g., test rules) could not directly depend on the outputs.
Fixed so rules produce outputs
rule loads image, runs it to generate configs, then removes image
outputs are produced as a tar file in the outputs directory for the rule (reference-able by other rules)
test rule updated to simplify usage

### CODE REVIEW

File: rules/docker_config.bzl [hunk 1]
```diff
@@ -154,6 +157,10 @@ def _docker_toolchain_autoconfig_impl(ctx):
     if ctx.attr.git_repo:
         clone_repo_cmd = ("cd " + bazel_config_dir + " && git clone " +
                           ctx.attr.git_repo + " " + project_repo_dir)
+    if ctx.attr.repo_pkg_tar:
+        # if package tar was used then the command should expand it
+        clone_repo_cmd = ("mkdir %s/%s && tar -xf /%s -C %s/%s " % (bazel_config_dir, project_repo_dir, ctx.file.repo_pkg_tar.basename, bazel_config_dir, project_repo_dir))
+    print(clone_repo_cmd)
```
Comment (xingao267) [inline rules/docker_config.bzl]: remove this print?

File: rules/docker_config.bzl [hunk 2]
```diff
@@ -241,81 +239,60 @@ def _docker_toolchain_autoconfig_impl(ctx):
...
+        "_extract_tpl": attr.label(
+            default = Label("@base_images_docker//util:extract.sh.tpl"),
+            allow_files = True,
+            single_file = True,
+        ),
+        "_image_id_extractor": attr.label(
+            default = "@io_bazel_rules_docker//contrib:extract_image_id.py",
+            allow_files = True,
+            single_file = True,
+        ),
```
Comment (xingao267) [inline rules/docker_config.bzl]: attrs = _container.image.attrs + _extract.attrs + {...}

and remove these

File: rules/docker_config.bzl [hunk 3]
```diff
@@ -241,81 +239,60 @@ def _docker_toolchain_autoconfig_impl(ctx):
...
+    # Include the repo_pkg_tar if needed
+    files = [install_sh] + ctx.files._installers
     if ctx.attr.repo_pkg_tar:
-        runfiles_list += ctx.files.repo_pkg_tar
-
-    runfiles = ctx.runfiles(files = runfiles_list)
+        files += [ctx.file.repo_pkg_tar]
+
+    image_tar = ctx.new_file(ctx.attr.name + ".tar")
+    load_image_sh_file = ctx.new_file("load.sh")
```
Comment (xingao267) [inline rules/docker_config.bzl]: is this used?

File: rules/docker_config.bzl [hunk 4]
```diff
@@ -241,81 +239,60 @@ def _docker_toolchain_autoconfig_impl(ctx):
...
+            single_file = True,
+        ),
+        "_image_id_extractor": attr.label(
+            default = "@io_bazel_rules_docker//contrib:extract_image_id.py",
+            allow_files = True,
+            single_file = True,
+        ),
     },
-    executable = True,
     outputs = _container.image.outputs + {
```
Comment (xingao267) [inline rules/docker_config.bzl]: not sure if this is possible, can we not expose _container.image.outputs?

File: rules/docker_config.bzl [hunk 5]
```diff
@@ -193,11 +194,11 @@ def _docker_toolchain_autoconfig_impl(ctx):
     copy_cmd = []
     for config_repo in ctx.attr.config_repos:
         src_dir = "$(bazel info output_base)/" + _EXTERNAL_FOLDER_PREFIX + config_repo
-        copy_cmd.append("cp -dr " + src_dir + " " + bazel_config_dir + "/")
+        copy_cmd.append("cp -dr " + src_dir + " " + "/")
 
         # We need to change the owner of the files we copied, so that they can
         # be manipulated from outside the container.
-        copy_cmd.append("chown -R $USER_ID " + bazel_config_dir + "/" + config_repo)
+        copy_cmd.append("tar -cf /outputs.tar " + " /".join(ctx.attr.config_repos))
```
Comment (xingao267) [inline rules/docker_config.bzl]: Update comments in new line 192 and 201

File: rules/docker_config.bzl [hunk 6]
```diff
@@ -160,6 +157,10 @@ def _docker_toolchain_autoconfig_impl(ctx):
     if ctx.attr.git_repo:
         clone_repo_cmd = ("cd " + bazel_config_dir + " && git clone " +
                           ctx.attr.git_repo + " " + project_repo_dir)
+    if ctx.attr.repo_pkg_tar:
+        # if package tar was used then the command should expand it
+        clone_repo_cmd = ("mkdir %s/%s && tar -xf /%s -C %s/%s " % (bazel_config_dir, project_repo_dir, ctx.file.repo_pkg_tar.basename, bazel_config_dir, project_repo_dir))
```
Comment (xingao267) [inline rules/docker_config.bzl]: wrap line

File: rules/docker_config.bzl [hunk 7]
```diff
@@ -241,81 +239,60 @@ def _docker_toolchain_autoconfig_impl(ctx):
...
+        image = image_tar,
+        commands = ["/" + ctx.attr.name + "_install.sh"],
+        extract_file = "/output.tar",
+        output_file = ctx.outputs.output_tar,
     )
 
 docker_toolchain_autoconfig_ = rule(
     attrs = _container.image.attrs + {
         "config_repos": attr.string_list(default = ["local_config_cc"]),
         "use_default_project": attr.bool(default = False),
```
Comment (xingao267) [inline rules/docker_config.bzl]: Can we rely on if repo_pkg_tar is specified to tell use_default_project or not?

File: rules/docker_config.bzl [hunk 8]
```diff
@@ -160,6 +157,10 @@ def _docker_toolchain_autoconfig_impl(ctx):
     if ctx.attr.git_repo:
         clone_repo_cmd = ("cd " + bazel_config_dir + " && git clone " +
                           ctx.attr.git_repo + " " + project_repo_dir)
+    if ctx.attr.repo_pkg_tar:
```
Comment (xingao267) [inline rules/docker_config.bzl]: repo_pkg_tar is only set when git_repo is none, maybe 'else' works better here.

File: tests/config/autoconfig_test.sh [hunk 9]
```diff
@@ -28,39 +28,9 @@ set -e
 WORKSPACE_ROOT=$(pwd)
```
Comment (xingao267) [inline tests/config/autoconfig_test.sh]: debian8_clang_autoconfig_test.sh can probably be simplified as well.

File: rules/docker_config.bzl (hunk 1 above)
Comment (nlopezgi) [reply rules/docker_config.bzl]: done

File: rules/docker_config.bzl (hunk 5 above)
Comment (nlopezgi) [reply rules/docker_config.bzl]: done

**Pull request merged**